src/java/org/apache/cassandra/cql3/constraints/ScalarColumnConstraint.java

```diff
@@ -0,0 +1,167 @@
+        {
+            this.param = param;
+            this.relationType = relationType;
+            this.term = term;
+        }
+
+        public ScalarColumnConstraint prepare()
+        {
+            return new ScalarColumnConstraint(param, relationType, term);
+        }
+    }
+
+    private ScalarColumnConstraint(ColumnIdentifier param, Operator relationType, String term)
+    {
+        this.param = param;
+        this.relationType = relationType;
+        this.term = term;
+    }
+
+    @Override
+    public void evaluate(Class<? extends AbstractType> valueType, Object columnValue)
+    {
+        Number columnValueNumber;
+        float sizeConstraint;
+
+        try
+        {
+            columnValueNumber = (Number) columnValue;
+            sizeConstraint = Float.parseFloat(term);
+        }
+        catch (NumberFormatException exception)
+        {
+            throw new ConstraintViolationException(param + " and " + term + " need to be numbers.");
+        }
+
+        ByteBuffer buffera = ByteBufferUtil.bytes(columnValueNumber.floatValue());
+        ByteBuffer bufferb = ByteBufferUtil.bytes(sizeConstraint);
+
+        if (!relationType.isSatisfiedBy(FloatType.instance, buffera, bufferb))
+            throw new ConstraintViolationException(columnValueNumber + " does not satisfy length constraint. "
+                                                   + sizeConstraint + " should be " + relationType + ' ' + term);
+    }
+
+    @Override
+    public void validate(ColumnMetadata columnMetadata) throws InvalidConstraintDefinitionException
+    {
+        if (!columnMetadata.type.isNumber())
+            throw new InvalidConstraintDefinitionException(param + " is not a number");
+    }
+
+    @Override
+    public ConstraintType getConstraintType()
+    {
+        return ConstraintType.SCALAR;
+    }
+
+    @Override
+    public String toString()
+    {
+        return param + " " + relationType + " " + term;
+    }
+
+    @Override
+    public IVersionedSerializer<ScalarColumnConstraint> serializer()
+    {
+        return serializer;
+    }
+
+    @Override
+    public void appendCqlTo(CqlBuilder builder)
+    {
+        builder.append(toString());
+    }
+
+    private static class Serializer implements IVersionedSerializer<ScalarColumnConstraint>
+    {
+        @Override
+        public void serialize(ScalarColumnConstraint columnConstraint, DataOutputPlus out, int version) throws IOException
+        {
+            out.writeUTF(columnConstraint.param.toString());
+            out.writeUTF(columnConstraint.relationType.toString());
+            out.writeUTF(columnConstraint.term);
+        }
+
+        @Override
+        public ScalarColumnConstraint deserialize(DataInputPlus in, int version) throws IOException
+        {
+            ColumnIdentifier param = new ColumnIdentifier(in.readUTF(), true);
+            Operator relationType = Operator.valueOf(in.readUTF());
```

Comment:
`serialize` writes the `toString` which usually does not return the same string as `name()`, making these serializations non-roundtrippable. This should probably use the `b` int field on `Operator` like `RowFilter.Serializer` does. Or even better, something similar but which would allow us to write a VInt rather than an int.

Reply:
I found this issue by triggering a metadata snapshot using `nodetool cms snapshot` which serializes the current ClusterMetadata, including schema. A subsequent restart then reads the most recent snapshot when replaying the metadata log during startup, which fails currently. Another way to do this would be too add some constraints to the schema in `org.apache.cassandra.distributed.test.log.SnapshotTest`.